### DISCUSSION TITLE
LSP previous and next reference problem

### DISCUSSION THREAD

**Discussion by chronon** [LazyVim/LazyVim#4966] [Q&A] [answered]
When editing a PHP file where [intelephense](https://github.com/bmewburn/vscode-intelephense) is enabled as the LSP server, I'm running into a weird issue that breaks the `[[` and `]]` mappings. With one buffer open the prev/next reference mappings work as expected and which-key looks correct:

![wk-good](https://github.com/user-attachments/assets/52e35d2d-92c1-45da-8a9e-1be8694366cc)

However, if I open a second buffer and try either `[[` or `]]`, the cursor is not jumping to the prev/next reference but to the start of the containing method. The which-key hint shows this:

![wk-bad](https://github.com/user-attachments/assets/60b19263-79e3-4ac4-ae98-ebfc24756cb1)

If I `LspRestart intelephense` the prev/next mappings work for all open buffers...until I open a new buffer which reverts all buffers back to not working.

The problem is only with PHP files and intelephense, other filetypes/LSPs work as expected for me. Also, this used to work fine, so either a recent intelephense or LazyVim update changed things. I'm not sure where to open a bug report, but my guess is intelephense. Anyone else run into this problem?

**Comment by folke** [comment]
Might be a Neovim builtin ftplugin that sets those?
You can check with `:verbose nmap [`

**Reply by dpetka2001** [reply]
Indeed https://github.com/neovim/neovim/blob/fac96b72a59e15d483e03bdc43b9ba08ce6c828e/runtime/ftplugin/php.vim#L129-L150

So, I guess `vim.g.no_php_maps = true` should disable it?

**Reply by folke** [reply]
Urgh, I really don't like it when those ftplugins change things like that...

**Reply by chronon** [reply]
When setting `vim.g.no_php_maps = true`, `[[` works in the first buffer like before, but in any newly opened buffers after that the mapping is removed.

**Reply by dpetka2001** [reply]
I'm not good with vimscript, so I have a hard time understanding what that does, but it seems to set `undo_ftplugin` to unmap `]]/[[` locally in the buffer. I would assume that setting `vim.g.no_php_maps = true` would prevent that `if` clause from happening. Can you also try `vim.g.no_plugin_maps = true` additionally? Other than that, I'm afraid I can't be of further help.

**Reply by chronon** [reply]
Thanks @dpetka2001, I appreciate your suggestions. Unfortunately setting `vim.g.no_plugin_maps = true` gives the same results as `vim.g.no_php_maps = true` as described in my previous post.

**Reply by dpetka2001** [reply]
When you open another buffer what filetype is that buffer? Is it `php`? Could you also check in the other buffer what does `:=vim.g.no_php_maps` show? I believe you put the option in your `~/.config/nvim/lua/config/options.lua` right?

**Reply by chronon** [reply]
Yes, the other buffer is `php` filetype and I'm adding the option in `options.lua`. When the option is set both buffers are `true` for `:=vim.g.no_php_maps`. With the option removed both buffers are `nil`.

**Reply by dpetka2001** [reply]
Can you also try the following in your `~/.config/nvim/lua/config/autocmds.lua`
```lua
vim.api.nvim_create_autocmd("FileType", {
  pattern = "php",
  callback = function()
    vim.g.no_php_maps = 1
  end,
})
```
and see if that makes any difference?

**Reply by chronon** [reply]
Same results as setting it in `options.lua`, the mapping is present and working in the first buffer, but missing in the second buffer.

**Reply by dpetka2001** [reply]
Is the second buffer pure php or some blade file or something else? What does `set ft?` show?

**Reply by chronon** [reply]
The files I've been testing with are pure php, both say `filetype=php`.

**Reply by chronon** [reply]
If the second buffer I open is another filetype besides php (like Typescript) with a different LSP, both work as expected.

**Reply by dpetka2001** [reply]
How are you opening the files? Which fuzzy finder do you use?

**Reply by dpetka2001** [reply]
So, @chronon here's some things that I've observed. I use `fzf-lua` and observed the same thing you did even with `vim.g.no_php_maps = true`. But when I open 2 php files with `nvim file1.php file2.php`, then the mappings work fine on both buffers. If I open Neovim and restore the session where both files are open in buffers, then the mappings work as expected. When I open Neovim with one file and then open another one using Neotree, then the mappings work as expected on both buffers. I even tried switching from `fzf-lua` to `telescope` and there it also worked well. 

So, from my testing I've observed that when opening a file through `FzfLua files` picker, then for some reason the `undo_ftplugin` gets triggered and that deletes the mappings in the buffers that are opened through it. I'm not 100% sure that is what happens, but from my observations this is the explanation I've come up with. 

@folke Do you happen to know anything about what I've mentioned even if it is some inkling? Really sorry for pinging you. I know that from my observations this should better be asked in FzfLua repo, but I'll let OP do that if he wants to further proceed with this. I'm just curious as to why this would happen that why I pinged you, not expecting you to find an answer to this problem.

**Reply by dpetka2001** [reply]
Ended up opening a discussion in FzfLua repo. I'm really curious as to why this happens. Hopefully, FzfLua maintainer will have some more insight.

**Reply by chronon** [reply]
Excellent detective work @dpetka2001! I also use `fzf-lua` and can confirm opening buffers any other way besides the `fzf-lua` file picker resolves the issue. Thanks for opening the [discussion](https://github.com/ibhagwan/fzf-lua/discussions/1555) in the `fzf-lua` repo as well.

**Reply by dpetka2001** [reply]
Yeah i don't know why this happens. Hopefully when Fzflua maintainer is available again, he may have some idea or not. Because that global variable is supposed to not trigger those mappings, but somehow with FzfLua picker `undo_ftplugin` is getting triggered and the mappings get deleted.

**Reply by ibhagwan** [reply]
I gave my 2 cents in the subject in https://github.com/ibhagwan/fzf-lua/discussions/1555, this is totally normal, php file is opened, ftplugin runtime is sourced and buffer mappings are set unless `no_php_maps` is present, 

I have no idea why this doesn’t happen with Telelscope as the ftplugin should be triggered by it as well (this is correct behavior), I’m thinking this has something to do with the ordering of which mappings are executed first LazyVim LSP or the runtime?

In any event this isn’t a bug or something I can control, fortunately the workaround is simple, add this to your config:
```lua
vim.g.no_php_maps = 1
```

There could be a discussion on whether neovim ftplugin should overwrite existing buffer mappings (if such exists) but that discussion is for the upstream.

**Comment by dpetka2001** [comment]
@folke I believe I found where things go wrong, but I can't understand why.

Here are my observations:
- I started by adding a `print("LspAttach bufnr: ", args.buf)` here https://github.com/LazyVim/LazyVim/blob/2d7a04be47b94010fd6c1a2be0283c0d4e5738ae/lua/lazyvim/util/lsp.lua#L26-L36 here on line 29
- Then I added 
```lua
    if keys.id == "]]" then
      print("has: ", has)
    end
    local cond = not (keys.cond == false or ((type(keys.cond) == "function") and not keys.cond()))
    if keys.id == "]]" then
      print("keys.cond: ", keys.cond == false)
      print("cond: ", cond)
    end
```
on these lines here https://github.com/LazyVim/LazyVim/blob/2d7a04be47b94010fd6c1a2be0283c0d4e5738ae/lua/lazyvim/plugins/lsp/keymaps.lua#L85-L86
I noticed from this print that `cond = false` for the keys. In fact every key that has `cond` defined in the mapping based on `Snacks.words.is_enabled()` returns `false`
- I added a `print("mode: ", mode)` here https://github.com/folke/snacks.nvim/blob/b006f50278e0c6e47d5f02a3295ea699d7120aea/lua/snacks/words.lua#L84-L95 on line 94 and I noticed the following.
```lua

LspAttach bufnr:  1
mode:  n
mode:  n
mode:  n
mode:  n
LspAttach bufnr:  1
LspAttach bufnr:  1
LspAttach bufnr:  1
LspAttach bufnr:  1
mode:  n
mode:  n
mode:  n
mode:  n
LspAttach bufnr:  1
mode:  n
mode:  n
mode:  n
mode:  n
LspAttach bufnr:  1
LspAttach bufnr:  1
LspAttach bufnr:  1
LspAttach bufnr:  1
mode:  n
mode:  n
mode:  n
mode:  n
mode:  c
mode:  c
mode:  n
test
mode:  n
mode:  n
mode:  t
mode:  n
LspAttach bufnr:  18
mode:  t
mode:  t
mode:  t
mode:  t
LspAttach bufnr:  18
LspAttach bufnr:  18
LspAttach bufnr:  18
LspAttach bufnr:  18
LspAttach bufnr:  18
mode:  t
mode:  t
mode:  t
mode:  t
LspAttach bufnr:  18
LspAttach bufnr:  18
LspAttach bufnr:  18
LspAttach bufnr:  18
mode:  t
mode:  n
mode:  n
mode:  c
mode:  c
mode:  n
```
The second buffer with `bufnr 18` which is being opened by fzf-lua, the `LspAttach` event happens still in terminal mode (`mode: t` in the prints). Thus, because that mode is not `config.modes` for Snacks.words the condition returns `false` and that is why the key does not get created at all.

I have no idea why this happens or who is the main culprit here. fzf-lua or snacks.words?? 

Anyway these are just my observations. I hope they might help you come up with who the main culprit is and hopefully solve this. By the way this happens with Lua files as well. It doesn't matter the filetype. The `LspAttach` event for getting the lsp keymaps is being triggered in terminal mode when you open a file with fzf-lua.

**Comment by folke** [comment, marked as answer]
@dpetka2001 thank you for looking into this! Fixed in snacls

**Reply by chronon** [reply]
Thank you @dpetka2001 for really digging into this, and thank you @folke for the fix. I can confirm the issue is resolved after updating.
